src/spatial/nsm.jl

```diff
@@ -0,0 +1,269 @@
+end
+
+######################## SSA specific helper routines ########################
+"""
+reevaluate all rates, recalculate tentative site firing times, and reinit the priority queue
+"""
+function fill_rates_and_get_times!(aggregation::NSMJumpAggregation, u, t)
+    @unpack ma_jumps, diffusion_constants, spatial_system = aggregation
+
+    num_majumps = get_num_majumps(ma_jumps)
+    num_species = length(@view u[:,1]) #NOTE assumes u is a matrix with ith column being the ith site
+    num_sites = number_of_sites(spatial_system)
+    cur_rates = SpatialRates(num_majumps,num_species,num_sites)
+
+    @assert cur_rates.reaction_rates_sum == zeros(typeof(cur_rates.reaction_rates_sum[1]),num_sites)
+    @assert cur_rates.diffusion_rates_sum == zeros(typeof(cur_rates.diffusion_rates_sum[1]),num_sites)
+
+    pqdata = Vector{typeof(t)}(undef, num_sites)
+    for site in 1:num_sites
+        update_reaction_rates!(cur_rates, 1:num_majumps, u, ma_jumps, site)
+        update_diffusion_rates!(cur_rates, 1:num_species, diffusion_constants, u, site, spatial_system)
+        pqdata[site] = t + randexp(aggregation.rng) / get_site_rate(cur_rates, site)
+    end
+
+    aggregation.cur_rates = cur_rates
+    aggregation.pq = MutableBinaryMinHeap(pqdata)
+    nothing
+end
+
+"""
+    update_dependent_rates_and_firing_times!(p, u, t)
+
+recalculate jump rates for jumps that depend on the just executed jump (p.prev_jump)
+"""
+function update_dependent_rates_and_firing_times!(p, u, t)
+    jump = p.prev_jump
+    if is_diffusion(p, jump)
+        source_site = jump.site
+        target_site = jump.target_site
+        update_rates_after_diffusion!(p, u, t, source_site, target_site, jump.index)
+        for site in [source_site, target_site]
+            update_site_time!(p.pq, p.rng, p.cur_rates, site, t)
+        end
+    else
+        site = jump.site
+        update_rates_after_reaction!(p, u, t, site, reaction_id_from_jump(p,jump))
+        update_site_time!(p.pq, p.rng, p.cur_rates, site, t)
+    end
+end
+
+function update_site_time!(pq, rng, cur_rates, site, t)
+    site_rate = get_site_rate(cur_rates, site)
+    if site_rate > zero(typeof(site_rate))
+        update!(pq, site, t + randexp(rng) / site_rate)
+    else
+        update!(pq, site, typemax(t))
+    end
+end
+
+######################## helper routines for all spatial SSAs ########################
+function update_rates_after_reaction!(p, u, t, site, reaction_id)
+    update_reaction_rates!(p.cur_rates, p.dep_gr[reaction_id], u, p.ma_jumps, site)
+    update_diffusion_rates!(p.cur_rates, p.jumptovars_map[reaction_id], p.diffusion_constants, u, site, p.spatial_system)
+end
+
+function update_rates_after_diffusion!(p, u, t, source_site, target_site, species)
+    for site in [source_site, target_site]
+        update_reaction_rates!(p.cur_rates, p.vartojumps_map[species], u, p.ma_jumps, site)
+        update_diffusion_rates!(p.cur_rates, [species], p.diffusion_constants, u, site, p.spatial_system)
```

Comment:
Why not make a version that takes a scalar for `species` and avoid creating a vector? Or else, can it work with a tuple instead?

Reply:
good idea. Making a version that takes a scalar